stan/math/prim/core/complex_base.hpp

```diff
@@ -0,0 +1,275 @@
+#ifndef STAN_MATH_PRIM_CORE_COMPLEX_BASE_HPP
+#define STAN_MATH_PRIM_CORE_COMPLEX_BASE_HPP
+
+#include <stan/math/prim/fun/square.hpp>
+#include <complex>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Base class for complex numbers.  The template parameter `V` is the
+ * value type.  This class is used like the curiously recursive
+ * template pattern (CRTP) with derived type `complex<V>`.  Thus any
+ * extending class must be of the form `complex<V>`.
+ *
+ * @tparam V value type for extending complex class
+ */
+template <typename V>
+class complex_base {
+ protected:
+  /**
+   * Real part.
+   */
+  V re_;
+
+  /**
+   * Imaginary part.
+   */
+  V im_;
+
+ public:
+  /**
+   * Type of real and imaginary parts.
+   */
+  using value_type = V;
+
+  /**
+   * Derived complex type used for function return types.
+   */
+  using complex_type = std::complex<value_type>;
+
+  /**
+   * Construct a complex base with the specified real and imaginary
+   * parts.
+   *
+   * @tparam T real type (must be assignable to `V`)
+   * @tparam U imaginary type (must be assignable to `V`)
+   * @param[in] re real part
+   * @param[in] im imaginary part (default 0)
+   */
+  template <typename T, typename U>
+  complex_base(const T& re, const U& im) : re_(re), im_(im) {}
```

Comment:
Should these be passed by value or reference?

Reply:
By reference.  They might be primitive, `var`, or `fvar`.   Passing by value would be performant for primitives and `var`, but `fvar<double>` and `fvar<var>` are 16 bytes, and `fvar<fvar<double>>` and `fvar<fvar<var>>` are 32 bytes.

I could break out an arithmetic type or `var` base class that would deal with the 8-byte ones separately.  But I don't think there's a huge saving from passing primitives by value vs. constant reference and would rather not have the complication in the constructors that would be required to do all that type splitting on primitive or var vs. fvar.

Reply:
> By reference. They might be primitive, var, or fvar. Passing by value would be performant for primitives and var, but fvar<double> and fvar<var> are 16 bytes, and fvar<fvar<double>> and fvar<fvar<var>> are 32 bytes.

I've been flip flopping on this too much to make a declarative statement, but I think it comes down to cache locality whether it's good to pass by value or reference. Most modern processors have some KB size of L1 cache with 64 byte cache lines. In section 7.2.1 of the 2nd edition of the C++ templates book he comments that if a signature with a const& argument is not inlined then the compiler can end up clearing cache out

> However, passing an address can create uncertainties for the compiler when it compiles the caller's code: What is the callee doing with that address? In theory, the callee can change all the values that are "reachable" through that address. That means, that the compiler has to assume that all the values it may have cached (usually, in machine registers) are invalid after the call. Reloading all those values can be quite expensive. You may be thinking that we are passing by _constant_ reference: Cannot the compiler deduce from that that no change can happen? Unfortunately, that is not the case because the caller may modify the referenced object through it own, non-const reference.

And if the compiler can't deduce that none of the methods a method calls uses const_cast then I think the cache is almost guaranteed to be cleared out

> I could break out an arithmetic type or var base class that would deal with the 8-byte ones separately. But I don't think there's a huge saving from passing primitives by value vs. constant reference and would rather not have the complication in the constructors that would be required to do all that type splitting on primitive or var vs. fvar.

idt that's necessary right now. lemme leave this open but consider this optional unless I have time to sit down and do enough tests to have some declarative statement about performance

Reply:
> Most modern processors have some KB size of L1 cache

Also relevant to this, if the machine can make those copies in cache then it can cost a few less cycles, Dan Lemire has a [blog post](https://lemire.me/blog/2019/09/05/passing-integers-by-reference-can-be-expensive/) with an example of that but I'm not sure if the example he uses matches up here (or generally)

Reply:
If we can see this in profiling, then it's definitely time to change the thousands of `const var&`  and `const fvar<T>&` args everywhere.

Reply:
Had some time to sit down today and workout a benchmark for this. The file [here](https://github.com/seantalts/perf-math/blob/test/eigen_pf/constref_move_copy_var.cpp) is a google benchmark that tries to just isolate the cost of moving by const ref, perfect forwarding, and by a straight copy. The `noinline` and `no_icf` in the attributes is to mock a large function that wouldn't be inlined and to stop any of the functions in there from getting merged together during the Identical Code Folding stage of most compilers.

Do you think the above example code fairly estimates the effect of passing arguments by certain types? Open to changes! Below is the results from running 

```bash
make constref_move_copy_var.o && make constref_move_copy_var
./constref_move_copy_var --benchmark_repetitions=50 --benchmark_report_aggregates_only=true
```

The `benchmark_repetitions` flag means to run the entire benchmark process N times. A single benchmark looks like this

```
----------------------------------------------------------
Benchmark                Time             CPU   Iterations
----------------------------------------------------------
add_const_bench       14.7 ns         14.7 ns     42042642
add_pf_bench          12.1 ns         12.1 ns    130300991
add_copy_bench        5.40 ns         5.40 ns    128300668
```

Where Benchmark is each individual test, time wall time, cpu is cpu time, and iterations is the number of iterations the benchmark was run. The `benchmark_repititions` flag does the above N times then takes the mean, median, and stddev

```
2020-03-03 12:12:29
Running ./constref_move_copy_var
Run on (32 X 3500 MHz CPU s)
CPU Caches:
  L1 Data 32K (x16)
  L1 Instruction 64K (x16)
  L2 Unified 512K (x16)
  L3 Unified 8192K (x4)
Load Average: 0.50, 0.42, 0.28
-----------------------------------------------------------------
Benchmark                       Time             CPU   Iterations
-----------------------------------------------------------------
add_const_bench_mean         5.57 ns         5.57 ns           50
add_const_bench_median       5.37 ns         5.37 ns           50
add_const_bench_stddev       1.31 ns         1.31 ns           50
add_pf_bench_mean            5.57 ns         5.57 ns           50
add_pf_bench_median          5.43 ns         5.43 ns           50
add_pf_bench_stddev         0.954 ns        0.954 ns           50
add_copy_bench_mean          5.31 ns         5.31 ns           50
add_copy_bench_median        5.25 ns         5.25 ns           50
add_copy_bench_stddev       0.359 ns        0.358 ns           50
```

Suprising! Doing values by copy seems to be about 4% faster! tbh I'm kind of suprised by that result, if you can find anything in the benchmark that makes the above seem fishy I'm open to any changes.

I haven't done a fresh install in a while but I think you should be able to copy and paste the below to run the benchmark

```bash
git clone --recursive https://github.com/seantalts/perf-math.git
git checkout eigen_pf
cd perf-math/benchmark
git clone https://github.com/google/googletest
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=RELEASE
make
cd ..
make constref_move_copy_var.o && make constref_move_copy_var
./constref_move_copy_var --benchmark_repetitions=50 --benchmark_report_aggregates_only=true
```

Reply:
(also fyi this is just a side convo if we change anything because of the above we should do it in a separate PR)

Reply:
Can you please say what's being tested here @SteveBronder.

And can we move this over to Discourse rather than having this long discussion via a pending issue?

Reply:
Yeah sure I'll close this and do something on discourse